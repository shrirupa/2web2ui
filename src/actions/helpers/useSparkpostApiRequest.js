import { useState, useEffect } from 'react';
import { sparkpost as sparkpostAxios } from 'src/helpers/axiosInstances';
import store from 'src/store';
import _ from 'lodash';

function useSparkpostApiRequest(options) {
  const [response, setResponse] = useState({});

  // Force effect to rerun by using a toggled boolean as a dependant
  const [retryToggle, setRetryToggle] = useState(false);
  const retry = () => setRetryToggle(!retryToggle);

  const { auth } = store.getState();
  const httpOptions = { ...options };

  if (auth.loggedIn) {
    _.set(httpOptions, 'headers.Authorization', auth.token);
  }

  useEffect(() => {
    setResponse({ pending: true, error: null });

    sparkpostAxios(httpOptions).then((res) => {
      const results = _.get(res, 'data.results', res.data);
      setResponse({ pending: false, results });
    }).catch((error) => {
      setResponse({ pending: false, error: error });
    });
  }, [retryToggle]);

  return { ...response, retry };
}

export default useSparkpostApiRequest;
