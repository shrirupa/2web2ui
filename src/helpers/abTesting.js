import _ from 'lodash';

/**
 * Formats edit mode's redux-form values to be passed on to the API
 */
export const formatFormValues = ({ default_template, variants, dates, ...rest }) => {
  let values = _.clone(rest);

  if (values.test_mode === 'learning') {
    values = _.omit(values, 'confidence_level');
  } else {
    values.confidence_level = parseFloat(values.confidence_level);
  }

  if (values.audience_selection === 'percent') {
    values = _.omit(values, 'total_sample_size');
    delete default_template.sample_size;
    variants = _.map(variants, (variant) => _.omit(variant, 'sample_size'));
  } else {
    values.total_sample_size = calculateTotalSampleSize({ default_template, variants });
    delete default_template.percent;
    variants = _.map(variants, (variant) => _.omit(variant, 'percent'));
  }

  // Cast stringified values to numbers
  default_template[values.audience_selection] = parseFloat(default_template[values.audience_selection]);
  _.forEach(variants, (variant) => variant[values.audience_selection] = parseFloat(variant[values.audience_selection]));
  values.engagement_timeout = parseInt(values.engagement_timeout, 10);

  return {
    ...values,
    start_time: dates.from,
    end_time: dates.to,
    default_template: reduceTemplateObject(default_template),
    variants: variants.map(reduceTemplateObject)
  };
};

/**
 * Formats template object (default template and variants) from redux-form to be passed on to the API
 * The template typeahead returns the full template object, we only need its ID
 */
export const reduceTemplateObject = ({ template_object, ...rest } = {}) => ({
  template_id: _.get(template_object, 'id'),
  ...rest
});

/**
 * Formats template object (default template and variants) from the API to be passed on to redux-form
 * The template typeahead expects full template objects
 */
export const findTemplateObject = (templates, { template_id, ...rest } = {}) => ({
  template_object: _.find(templates, ({ id }) => id === template_id),
  ...rest
});

/**
 * Calculates the total_sample_size of a test based upon the sample_size of default_template & variants
 */
export const calculateTotalSampleSize = ({ default_template, variants }) =>
  _.reduce(
    variants,
    (sum, variant) => sum + parseInt(variant.sample_size, 10),
    parseInt(default_template.sample_size, 10)
  );

export const formatDraftOptionalValues = (values) => {
  _.remove(values.variants, (variant) => 'percent' in variant && !_.isFinite(variant.percent));
  _.remove(values.variants, (variant) => 'sample_size' in variant && !_.isFinite(variant.sample_size));

  if (values.variants.length === 0) {
    delete values.variants;
  }

  if ('percent' in values.default_template && !_.isFinite(values.default_template.percent)) {
    values.default_template.percent = 100;
  }

  if ('sample_size' in values.default_template && !_.isFinite(values.default_template.sample_size)) {
    values.default_template.sample_size = 1;
  }

  if ('total_sample_size' in values && (values.total_sample_size === 0 || isNaN(values.total_sample_size))) {
    values.total_sample_size = 1;
  }

  return values;
};
