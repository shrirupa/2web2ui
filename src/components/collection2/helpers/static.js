import { objectSortMatch } from 'src/helpers/sortMatch';
const objectValuesToString = (keys) => (item) => (keys || Object.keys(item)).map((key) => item[key]).join(' ');

/*
 * Filter a fixed rowset using objectSortMatch()
 */
const filter = (rows, pattern = '', { keyMap, itemToStringKeys, matchThreshold } = {}) => {
  if (!pattern) {
    return { rows, rowCount: rows.length };
  }
  const filtered = objectSortMatch({
    items: rows,
    pattern,
    getter: objectValuesToString(itemToStringKeys),
    keyMap,
    matchThreshold
  });
  return {
    rows: filtered,
    rowCount: filtered.length
  };
};

export const filteringModel = (rows, options) => ({ pattern } = {}) => filter(rows, pattern, options);

/*
 * Construct a single "page" of results from a fixed rowset
 */
const paginate = (rows, currentPage = 1, perPage = 25) => {
  const currentIndex = (currentPage - 1) * perPage;
  return {
    rows: rows.slice(currentIndex, currentIndex + perPage),
    rowCount: rows.length
  };
};

export const paginatingModel = (rows) => ({ currentPage, perPage } = {}) => paginate(rows, currentPage, perPage);

/*
 * Create a filter/pagination function for a fixed list of records. This is meant for use
 * as a fetchRows implementation for a Collection.
 */
export const fullModel = (rows, { keyMap = {}, itemToStringKeys, matchThreshold } = {}) =>
  ({ pattern, currentPage, perPage } = {}) => {
    const { rows: filtered, rowCount } = filter(rows, pattern, { keyMap, itemToStringKeys, matchThreshold });
    return {
      rows: paginate(filtered, currentPage, perPage).rows,
      rowCount
    };
  };
