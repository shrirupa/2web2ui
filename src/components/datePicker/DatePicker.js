/* eslint-disable max-lines */
import React, { Component } from 'react';
import { subMonths, format } from 'date-fns';
import { getStartOfDay, getEndOfDay, getRelativeDateOptions, getNextHour, isSameDate } from 'src/helpers/date';
import { roundBoundaries } from 'src/helpers/metrics';
import { Button, TextField, Select, Popover, WindowEvent, Error } from '@sparkpost/matchbox';
import DateSelector from 'src/components/dateSelector/DateSelector';
import ManualEntryForm from './ManualEntryForm';
import { FORMATS } from 'src/constants';
import styles from './DatePicker.module.scss';
import PropTypes from 'prop-types';

export default class AppDatePicker extends Component {

  DATE_FORMAT = FORMATS.LONG_DATETIME;
  state = {
    showDatePicker: false,
    selecting: false,
    selected: { },
    validationError: null
  }

  componentDidMount() {
    this.syncTimeToState(this.props);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.from !== this.props.from || nextProps.to !== this.props.to) {
      this.syncTimeToState(nextProps);
    }
  }

  // Sets local state from reportOptions redux state - need to separate to handle pre-apply state
  syncTimeToState = ({ from, to }) => {
    this.setState({ selected: { from, to }});
  }

  // Closes popover on escape, submits on enter
  handleKeyDown = (e) => {
    if (!this.state.showDatePicker) {
      return;
    }

    if (e.key === 'Escape') {
      this.cancelDatePicker();
    }

    if (!this.state.selecting && e.key === 'Enter') {
      this.handleSubmit();
    }
  }

  handleDayKeyDown = (day, modifiers, e) => {
    this.handleKeyDown(e);
    e.stopPropagation();
  }

  cancelDatePicker = () => {
    this.syncTimeToState(this.props);
    this.setState({ showDatePicker: false });
  }

  showDatePicker = () => {
    this.setState({ showDatePicker: true });
  }

  handleDayClick = (clicked) => {
    const { selecting, selected } = this.state;
    const { validate } = this.props;

    const dates = selecting
      ? selected
      : { from: this.fromFormatter(clicked), to: getEndOfDay(clicked, { preventFuture: this.props.preventFuture }) };

    const validationError = validate && validate(dates);

    if (selecting && validationError) {
      this.setState({ validationError });
      return;
    }

    this.setState({
      selected: dates,
      beforeSelected: dates,
      selecting: !selecting,
      validationError: null
    });
  }

  handleDayHover = (hovered) => {
    const { selecting } = this.state;

    if (selecting) {
      this.setState({ selected: { ...this.getOrderedRange(hovered) }});
    }
  }

  getOrderedRange(newDate) {
    let { from, to } = this.state.beforeSelected;

    if (from.getTime() <= newDate.getTime()) {
      to = getEndOfDay(newDate, { preventFuture: this.props.preventFuture });
    } else {
      from = this.fromFormatter(newDate);
    }

    if (this.props.roundToPrecision) {
      const rounded = roundBoundaries(from, to);
      from = rounded.from.toDate();
      to = rounded.to.toDate();
    }
    return { from, to };
  }

  handleSelectRange = (e) => {
    const value = e.currentTarget.value;

    if (value === 'custom') {
      this.setState({ showDatePicker: true });
    } else {
      this.setState({ showDatePicker: false });
      this.props.onChange({ relativeRange: value });
    }
  }

  handleFormDates = ({ from, to }, callback) => {
    this.setState({ selected: { from, to }}, () => callback());
  }

  handleSubmit = () => {
    if (this.state.validationError) {
      return;
    }

    this.setState({ showDatePicker: false, selecting: false, validationError: null });
    this.props.onChange({ ...this.state.selected, relativeRange: 'custom' });
  }

  handleTextUpdate = () => {
    if (this.props.onBlur) {
      this.props.onBlur();
    }
  }

  fromFormatter = (fromDate) => {
    const isDateToday = isSameDate(getStartOfDay(fromDate), getStartOfDay(new Date()));
    const formatter = (this.props.fromSelectsNextHour && isDateToday) ? getNextHour : getStartOfDay;
    return formatter(fromDate);
  }

  render() {
    const { selected: { from, to }, showDatePicker, validationError } = this.state;
    const selectedRange = showDatePicker ? 'custom' : this.props.relativeRange;

    // allow for prop-level override of "now" (DI, etc.)
    const {
      now = new Date(),
      relativeDateOptions = [],
      disabled,
      datePickerProps = {},
      textFieldProps = {},
      dateFieldFormat,
      roundToPrecision,
      preventFuture,
      showPresets = true,
      error,
      left,
      hideManualEntry
    } = this.props;
    const dateFormat = dateFieldFormat || this.DATE_FORMAT;

    const rangeSelect = showPresets
      ? <Select
        options={getRelativeDateOptions(relativeDateOptions)}
        onChange={this.handleSelectRange}
        value={selectedRange}
        disabled={disabled} />
      : null;

    const dateField = <TextField
      onClick={this.showDatePicker}
      connectLeft={rangeSelect}
      value={`${format(from, dateFormat)} â€“ ${format(to, dateFormat)}`}
      readOnly
      onBlur={this.handleTextUpdate}
      error={error}
      disabled={disabled}
      {...textFieldProps} />;

    return (
      <Popover
        wrapper='div'
        className={styles.Popover}
        trigger={dateField}
        onClose={this.cancelDatePicker}
        open={this.state.showDatePicker}
        left={left} >

        <DateSelector
          numberOfMonths={2}
          fixedWeeks
          enableOutsideDays={false}
          initialMonth={subMonths(now, 1)}
          toMonth={now}
          disabledDays={{ after: now }}
          onDayClick={this.handleDayClick}
          onDayMouseEnter={this.handleDayHover}
          onDayFocus={this.handleDayHover}
          onKeyDown={this.handleKeyDown}
          onDayKeyDown={this.handleDayKeyDown}
          selectedDays={this.state.selected}
          {...datePickerProps}
        />

        {!hideManualEntry && (
          <ManualEntryForm
            selectDates={this.handleFormDates}
            onEnter={this.handleKeyDown}
            to={to}
            from={from}
            roundToPrecision={roundToPrecision}
            preventFuture={preventFuture}
          />
        )}

        <Button primary onClick={this.handleSubmit} className={styles.Apply}>Apply</Button>
        <Button onClick={this.cancelDatePicker}>Cancel</Button>
        {validationError && (
          <span className={styles.Error}>
            <Error wrapper='span' error={validationError}></Error>
          </span>
        )}
        <WindowEvent event='keydown' handler={this.handleKeyDown} />
      </Popover>
    );
  }
}

AppDatePicker.propTypes = {
  now: PropTypes.instanceOf(Date),
  from: PropTypes.instanceOf(Date),
  to: PropTypes.instanceOf(Date),
  relativeRange: PropTypes.string,
  relativeDateOptions: PropTypes.arrayOf(PropTypes.string),
  roundToPrecision: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  datePickerProps: PropTypes.object,
  dateFieldFormat: PropTypes.string,
  disabled: PropTypes.bool,
  showPresets: PropTypes.bool,
  hideManualEntry: PropTypes.bool
};

AppDatePicker.defaultProps = {
  roundToPrecision: true,
  preventFuture: true
};
