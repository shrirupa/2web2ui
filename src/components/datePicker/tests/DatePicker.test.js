import React from 'react';
import { shallow } from 'enzyme';
import _ from 'lodash';
import * as dateHelpers from 'src/helpers/date';
import * as metricsHelpers from 'src/helpers/metrics';
import datefns from 'date-fns';
import moment from 'moment';
import DatePicker from '../DatePicker';
import utc from 'src/__testHelpers__/time';

jest.mock('src/helpers/date');
jest.mock('react-dom');
jest.mock('date-fns');

describe('Component: DatePicker', () => {

  let wrapper;
  let instance;
  let props;
  let mockNow;
  let mockFrom;

  beforeEach(() => {
    mockNow = utc({ year: 2017, month: 1, day: 15, hour: 12 });
    mockFrom = utc({ year: 2017, month: 1, day: 15, hour: 11 });
    props = {
      from: mockFrom,
      to: mockNow,
      relativeRange: 'day',
      relativeDateOptions: [],
      onChange: jest.fn(),
      onBlur: jest.fn(),
      now: mockNow,
      disabled: false
    };

    dateHelpers.getStartOfDay = jest.fn(() => 'start-of-day');
    dateHelpers.getNextHour = jest.fn(() => 'next-hour');
    dateHelpers.getEndOfDay = jest.fn(() => 'end-of-day');
    dateHelpers.isSameDate = jest.fn(() => false);
    metricsHelpers.roundBoundaries = jest.fn(() => ({ from: moment(mockFrom), to: moment(mockNow) }));
    dateHelpers.getRelativeDateOptions = jest.fn(() => [1, 2, 3]);
    datefns.format = jest.fn((a,b) => b);
    datefns.subMonths = jest.fn((a) => a);

    wrapper = shallow(<DatePicker {...props} />);
    instance = wrapper.instance();

    // spy on all instance methods
    _.functions(instance).forEach((f) => jest.spyOn(instance, f));
  });

  it('should render ok by default', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render range select if showPresets is false', () => {
    wrapper.setProps({ showPresets: false });
    expect(wrapper.find('Popover').prop('trigger')).toMatchSnapshot();
  });

  it('should properly mount', () => {
    jest.spyOn(window, 'addEventListener');

    instance.componentDidMount();
    expect(instance.syncTimeToState).toHaveBeenCalledTimes(1);
  });

  it('should not render form fields when manual entry is hidden', () => {
    wrapper.setProps({ hideManualEntry: true });
    expect(wrapper.find('ManualEntryForm')).not.toExist();
  });

  describe('syncTimeToState', () => {

    const before = { from: 'unchanged', to: 'unchanged' };
    const after = { from: new Date(), to: new Date() };

    it('should sync the state when from or to is changed', () => {
      wrapper.setState({ selected: before });
      expect(wrapper.state('selected')).toEqual(before);
      wrapper.setProps(after);
      expect(instance.syncTimeToState).toHaveBeenCalledTimes(1);
      expect(wrapper.state('selected')).toEqual(after);
    });

    it('shouldn\'t sync the state when from and to aren\'t changed', () => {
      wrapper.setState({ selected: before });
      wrapper.setProps({ other: 'stuff' });
      expect(instance.syncTimeToState).not.toHaveBeenCalled();
      expect(wrapper.state('selected')).toEqual(before);
    });

  });

  describe('handleKeyDown', () => {

    const ESCAPE = { key: 'Escape' };
    const ENTER = { key: 'Enter' };

    it('should do nothing if picker is closed already', () => {
      wrapper.setState({ showDatePicker: false });
      instance.handleKeyDown(ESCAPE);
      expect(instance.cancelDatePicker).not.toHaveBeenCalled();
      expect(instance.handleSubmit).not.toHaveBeenCalled();
      expect(wrapper.state('showDatePicker')).toEqual(false);
    });

    it('should cancel date picker on escape keydown', () => {
      wrapper.setState({ showDatePicker: true });
      instance.handleKeyDown(ESCAPE);
      expect(instance.cancelDatePicker).toHaveBeenCalledTimes(1);
      expect(instance.handleSubmit).not.toHaveBeenCalled();
      expect(wrapper.state('showDatePicker')).toEqual(false);
    });

    it('should submit on enter', () => {
      wrapper.setState({ showDatePicker: true, selecting: false });
      instance.handleKeyDown(ENTER);
      expect(instance.cancelDatePicker).not.toHaveBeenCalled();
      expect(instance.handleSubmit).toHaveBeenCalledTimes(1);
      expect(wrapper.state('showDatePicker')).toEqual(false);
    });

    it('should not submit if selecting is true', () => {
      wrapper.setState({ showDatePicker: true, selecting: true });
      instance.handleKeyDown(ENTER);
      expect(instance.cancelDatePicker).not.toHaveBeenCalled();
      expect(instance.handleSubmit).not.toHaveBeenCalled();
      expect(wrapper.state('showDatePicker')).toEqual(true);
    });

  });

  describe('handleDayKeyDown', () => {

    it('should stop propagation on a day key down event', () => {
      const e = { stopPropagation: jest.fn() };
      const x = '';
      instance.handleDayKeyDown(x, x, e);
      expect(instance.handleKeyDown).toHaveBeenCalledWith(e);
      expect(e.stopPropagation).toHaveBeenCalledTimes(1);
    });

  });

  describe('handleTextUpdate', () => {
    it('should call props.onBlur if it exists', () => {
      instance.handleTextUpdate();
      expect(props.onBlur).toHaveBeenCalled();
    });

    it('should do nothing if props.onBlur does not exist', () => {
      wrapper.setProps({ onBlur: undefined });
      instance.handleTextUpdate();
      expect(props.onBlur).not.toHaveBeenCalled();
    });
  });

  describe('cancelDatePicker', () => {

    it('should sync props and close date picker', () => {
      wrapper.setState({ showDatePicker: true });
      instance.cancelDatePicker();
      expect(instance.syncTimeToState).toHaveBeenCalledWith(instance.props);
      expect(wrapper.state('showDatePicker')).toEqual(false);
    });

  });

  describe('showDatePicker', () => {

    it('should set date picker state', () => {
      wrapper.setState({ showDatePicker: false });
      instance.showDatePicker();
      expect(wrapper.state('showDatePicker')).toEqual(true);
    });

  });

  describe('handleDayClick', () => {

    it('should handle a day click while selecting', () => {
      const mockSelected = {};
      wrapper.setState({ selecting: true, selected: mockSelected, beforeSelected: null });
      instance.handleDayClick();
      expect(dateHelpers.getStartOfDay).not.toHaveBeenCalled();
      expect(wrapper.state('selected')).toBe(mockSelected);
      expect(wrapper.state('beforeSelected')).toBe(mockSelected);
      expect(wrapper.state('selecting')).toEqual(false);
    });

    it('should handle a day click when not selecting (start of day)', () => {
      const mockSelected = {};
      const mockClicked = {};

      wrapper.setState({ selecting: false, selected: mockSelected, beforeSelected: null });
      const mockNewSelected = { from: 'start-of-day', to: 'end-of-day' };

      instance.handleDayClick(mockClicked);

      expect(dateHelpers.getStartOfDay).toHaveBeenCalledWith(mockClicked);
      expect(dateHelpers.getStartOfDay).toHaveBeenCalledTimes(3);
      expect(dateHelpers.getEndOfDay).toHaveBeenCalledWith(mockClicked, { preventFuture: true });
      expect(wrapper.state('selected')).toEqual(mockNewSelected);
      expect(wrapper.state('beforeSelected')).toEqual(mockNewSelected);
      expect(wrapper.state('selecting')).toEqual(true);
    });

    it('should handle a day click when not selecting (next hour)', () => {
      const mockSelected = {};
      const mockClicked = {};

      dateHelpers.isSameDate = jest.fn(() => true);

      wrapper.setProps({ fromSelectsNextHour: true });
      wrapper.setState({ selecting: false, selected: mockSelected, beforeSelected: null });
      const mockNewSelected = { from: 'next-hour', to: 'end-of-day' };

      instance.handleDayClick(mockClicked);

      expect(dateHelpers.getStartOfDay).toHaveBeenCalledWith(mockClicked);
      expect(dateHelpers.getStartOfDay).toHaveBeenCalledTimes(2);
      expect(dateHelpers.getNextHour).toHaveBeenCalledWith(mockClicked);
      expect(dateHelpers.getEndOfDay).toHaveBeenCalledWith(mockClicked, { preventFuture: true });
      expect(wrapper.state('selected')).toEqual(mockNewSelected);
      expect(wrapper.state('beforeSelected')).toEqual(mockNewSelected);
      expect(wrapper.state('selecting')).toEqual(true);
    });

    it('should handle a day click whith a validation error', () => {
      const validate = jest.fn(() => 'error oh no');
      const mockSelected = {};
      const mockClicked = {};

      dateHelpers.isSameDate = jest.fn(() => true);

      wrapper.setProps({ validate });
      wrapper.setState({ selecting: false, selected: mockSelected, beforeSelected: null });
      const mockNewSelected = { from: 'start-of-day', to: 'end-of-day' };

      instance.handleDayClick(mockClicked);
      expect(validate).toHaveBeenCalledWith(mockNewSelected);
      expect(wrapper.state('selecting')).toEqual(true);
    });

  });

  describe('handleDayHover', () => {

    it('should handle a day hover while selecting', () => {
      const mockOrderedRange = { from: 'from', to: 'to' };
      const mockHovered = {};
      instance.getOrderedRange = jest.fn(() => mockOrderedRange);
      wrapper.setState({ selecting: true });

      instance.handleDayHover(mockHovered);

      expect(instance.getOrderedRange).toHaveBeenCalledWith(mockHovered);
      expect(wrapper.state('selected')).toEqual(mockOrderedRange);
    });

    it('should handle a day hover when not selecting', () => {
      const mockHovered = {};
      const mockSelected = {};
      instance.getOrderedRange = jest.fn();
      wrapper.setState({ selecting: false, selected: mockSelected });

      instance.handleDayHover(mockHovered);

      expect(instance.getOrderedRange).not.toHaveBeenCalled();
      expect(wrapper.state('selected')).toEqual(mockSelected);
    });

  });

  describe('getOrderedRange', () => {

    it('should return correct range when new date is between from and to', () => {
      const from = new Date('2018-01-01');
      const newDate = new Date('2018-01-02');
      const to = new Date('2018-01-03');

      wrapper.setState({ beforeSelected: { from, to }});
      const range = instance.getOrderedRange(newDate);

      expect(dateHelpers.getEndOfDay).toHaveBeenCalledWith(newDate, { preventFuture: true });
      expect(dateHelpers.getStartOfDay).not.toHaveBeenCalled();
      expect(metricsHelpers.roundBoundaries).toHaveBeenCalledWith(from, 'end-of-day');
      expect(range).toEqual({ from: moment(mockFrom).toDate(), to: moment(mockNow).toDate() });
    });

    it('should return correct range when new date is before from and to (start of day)', () => {
      const newDate = new Date('2018-01-01');
      const from = new Date('2018-01-02');
      const to = new Date('2018-01-03');

      wrapper.setState({ beforeSelected: { from, to }});
      const range = instance.getOrderedRange(newDate);

      expect(dateHelpers.getEndOfDay).not.toHaveBeenCalled();
      expect(dateHelpers.getStartOfDay).toHaveBeenCalledWith(newDate);
      expect(dateHelpers.getStartOfDay).toHaveBeenCalledTimes(3);
      expect(metricsHelpers.roundBoundaries).toHaveBeenCalledWith('start-of-day', to);
      expect(range).toEqual({ from: moment(mockFrom).toDate(), to: moment(mockNow).toDate() });
    });

    it('should return correct range when new date is before from and to (next hour)', () => {
      const newDate = new Date('2018-01-01');
      const from = new Date('2018-01-02');
      const to = new Date('2018-01-03');

      dateHelpers.isSameDate = jest.fn(() => true);

      wrapper.setProps({ fromSelectsNextHour: true });
      wrapper.setState({ beforeSelected: { from, to }});
      const range = instance.getOrderedRange(newDate);

      expect(dateHelpers.getEndOfDay).not.toHaveBeenCalled();
      expect(dateHelpers.getNextHour).toHaveBeenCalledWith(newDate);
      expect(dateHelpers.getStartOfDay).toHaveBeenCalledTimes(2);
      expect(metricsHelpers.roundBoundaries).toHaveBeenCalledWith('next-hour', to);
      expect(range).toEqual({ from: moment(mockFrom).toDate(), to: moment(mockNow).toDate() });
    });

    it('should return correct range when new date is after from and to', () => {
      const from = new Date('2018-01-01');
      const to = new Date('2018-01-02');
      const newDate = new Date('2018-01-03');

      wrapper.setState({ beforeSelected: { from, to }});
      const range = instance.getOrderedRange(newDate);

      expect(dateHelpers.getEndOfDay).toHaveBeenCalledWith(newDate, { preventFuture: true });
      expect(dateHelpers.getStartOfDay).not.toHaveBeenCalled();
      expect(metricsHelpers.roundBoundaries).toHaveBeenCalledWith(from, 'end-of-day');
      expect(range).toEqual({ from: moment(mockFrom).toDate(), to: moment(mockNow).toDate() });
    });

    it('should return correct range when from and newDate are the same', () => {
      const from = new Date('2018-01-01');
      const newDate = new Date('2018-01-01');
      const to = new Date('2018-01-03');

      wrapper.setState({ beforeSelected: { from, to }});
      const range = instance.getOrderedRange(newDate);

      expect(dateHelpers.getEndOfDay).toHaveBeenCalledWith(newDate, { preventFuture: true });
      expect(dateHelpers.getStartOfDay).not.toHaveBeenCalled();
      expect(metricsHelpers.roundBoundaries).toHaveBeenCalledWith(from, 'end-of-day');
      expect(range).toEqual({ from: moment(mockFrom).toDate(), to: moment(mockNow).toDate() });
    });

  });

  describe('handleSelectRange', () => {
    let e;

    beforeEach(() => {
      e = {
        currentTarget: {}
      };
    });

    it('should open date picker for custom value', () => {
      e.currentTarget.value = 'custom';
      wrapper.setState({ showDatePicker: false });
      instance.handleSelectRange(e);
      expect(wrapper.state('showDatePicker')).toEqual(true);
      expect(props.onChange).not.toHaveBeenCalled();
    });

    it('should close date picker and refresh for non-custom value', () => {
      e.currentTarget.value = 'lasagna';
      wrapper.setState({ showDatePicker: true });
      instance.handleSelectRange(e);
      expect(wrapper.state('showDatePicker')).toEqual(false);
      expect(props.onChange).toHaveBeenCalledWith({ relativeRange: 'lasagna' });
    });

  });

  describe('handleFormDates', () => {

    it('should use a date field format if provided', () => {
      wrapper.setProps({ dateFieldFormat: 'YYYY-MM h' });
      expect(wrapper.find('Popover').props().trigger).toMatchSnapshot();
    });

    it('should set from and to state and call the passed in callback', () => {
      wrapper.setState({ selected: false });
      const mockCallback = jest.fn();
      const mockSelected = { from: 'from', to: 'to' };
      instance.handleFormDates(mockSelected, mockCallback);
      expect(wrapper.state('selected')).toEqual(mockSelected);
      expect(mockCallback).toHaveBeenCalledTimes(1);
    });

  });

  describe('handleSubmit', () => {

    it('should reset some state and refresh', () => {
      const mockSelected = { a: 1, b: 2, c: 3 };
      wrapper.setState({ showDatePicker: true, selecting: true, selected: mockSelected });
      instance.handleSubmit();
      expect(wrapper.state('showDatePicker')).toEqual(false);
      expect(wrapper.state('selecting')).toEqual(false);
      expect(props.onChange).toHaveBeenCalledWith({ ...mockSelected, relativeRange: 'custom' });
    });

    it('should not do anything with a validation error', () => {
      const mockSelected = { a: 1, b: 2, c: 3 };
      wrapper.setState({ showDatePicker: true, selecting: true, selected: mockSelected, validationError: 'oh no' });
      instance.handleSubmit();
      expect(wrapper.state('showDatePicker')).toEqual(true);
      expect(wrapper.state('selecting')).toEqual(true);
      expect(props.onChange).not.toHaveBeenCalled();
    });

  });

  describe('fromFormatter', () => {

    it('should format from date with getNextHour', () => {
      dateHelpers.isSameDate = jest.fn(() => true);
      wrapper.setProps({ fromSelectsNextHour: true });
      expect(instance.fromFormatter('from-date')).toEqual('next-hour');
      expect(dateHelpers.getNextHour).toHaveBeenCalledWith('from-date');
      expect(dateHelpers.getStartOfDay).toHaveBeenCalledTimes(2);
    });

    it('should format from date with getStartOfDay (fromSelectsNextHour === false)', () => {
      dateHelpers.isSameDate = jest.fn(() => true);
      expect(instance.fromFormatter('from-date')).toEqual('start-of-day');
      expect(dateHelpers.getNextHour).not.toHaveBeenCalled();
      expect(dateHelpers.getStartOfDay).toHaveBeenCalledWith('from-date');
      expect(dateHelpers.getStartOfDay).toHaveBeenCalledTimes(3);
    });

    it('should format from date with getStartOfDay (isDateToday === false)', () => {
      wrapper.setProps({ fromSelectsNextHour: true });
      expect(instance.fromFormatter('from-date')).toEqual('start-of-day');
      expect(dateHelpers.getNextHour).not.toHaveBeenCalled();
      expect(dateHelpers.getStartOfDay).toHaveBeenCalledWith('from-date');
      expect(dateHelpers.getStartOfDay).toHaveBeenCalledTimes(3);
    });
  });

});
