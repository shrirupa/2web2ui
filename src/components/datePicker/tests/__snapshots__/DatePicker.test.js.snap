// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: DatePicker handleFormDates should use a date field format if provided 1`] = `
<React.Fragment>
  <label
    className="Label"
    htmlFor="date-field-date-picker"
  >
    Narrow Date Range
  </label>
  <TextField
    connectLeft={
      <React.Fragment>
        <label
          className="Label"
          htmlFor="range-select-date-picker"
        >
          Broad Date Range
        </label>
        <Select
          disabled={false}
          id="range-select-date-picker"
          onChange={[MockFunction]}
          options={
            Array [
              1,
              2,
              3,
            ]
          }
          value="day"
        />
      </React.Fragment>
    }
    disabled={false}
    id="date-field-date-picker"
    onBlur={[MockFunction]}
    onClick={[MockFunction]}
    readOnly={true}
    resize="both"
    type="text"
    value="YYYY-MM h – YYYY-MM h"
  />
</React.Fragment>
`;

exports[`Component: DatePicker should not render range select if showPresets is false 1`] = `
<React.Fragment>
  <label
    className="Label"
    htmlFor="date-field-date-picker"
  >
    Narrow Date Range
  </label>
  <TextField
    connectLeft={null}
    disabled={false}
    id="date-field-date-picker"
    onBlur={[MockFunction]}
    onClick={[MockFunction]}
    readOnly={true}
    resize="both"
    type="text"
    value="MMM Do YYYY h:mma – MMM Do YYYY h:mma"
  />
</React.Fragment>
`;

exports[`Component: DatePicker should render ok by default 1`] = `
<Popover
  bottom={true}
  className="Popover"
  onClose={[Function]}
  open={false}
  right={true}
  trigger={
    <React.Fragment>
      <label
        className="Label"
        htmlFor="date-field-date-picker"
      >
        Narrow Date Range
      </label>
      <TextField
        connectLeft={
          <React.Fragment>
            <label
              className="Label"
              htmlFor="range-select-date-picker"
            >
              Broad Date Range
            </label>
            <Select
              disabled={false}
              id="range-select-date-picker"
              onChange={[Function]}
              options={
                Array [
                  1,
                  2,
                  3,
                ]
              }
              value="day"
            />
          </React.Fragment>
        }
        disabled={false}
        id="date-field-date-picker"
        onBlur={[Function]}
        onClick={[Function]}
        readOnly={true}
        resize="both"
        type="text"
        value="MMM Do YYYY h:mma – MMM Do YYYY h:mma"
      />
    </React.Fragment>
  }
  wrapper="div"
>
  <DateSelector
    disabledDays={
      Object {
        "after": 2017-01-15T12:00:00.000Z,
      }
    }
    enableOutsideDays={false}
    fixedWeeks={true}
    initialMonth={2017-01-15T12:00:00.000Z}
    numberOfMonths={2}
    onDayClick={[Function]}
    onDayFocus={[Function]}
    onDayKeyDown={[Function]}
    onDayMouseEnter={[Function]}
    onKeyDown={[Function]}
    selectedDays={
      Object {
        "from": 2017-01-15T11:00:00.000Z,
        "to": 2017-01-15T12:00:00.000Z,
      }
    }
    toMonth={2017-01-15T12:00:00.000Z}
  />
  <ManualEntryForm
    from={2017-01-15T11:00:00.000Z}
    onEnter={[Function]}
    preventFuture={true}
    roundToPrecision={true}
    selectDates={[Function]}
    to={2017-01-15T12:00:00.000Z}
  />
  <Button
    className="Apply"
    onClick={[Function]}
    primary={true}
    size="default"
  >
    Apply
  </Button>
  <Button
    onClick={[Function]}
    size="default"
  >
    Cancel
  </Button>
  <WindowEvent
    event="keydown"
    handler={[Function]}
  />
</Popover>
`;
